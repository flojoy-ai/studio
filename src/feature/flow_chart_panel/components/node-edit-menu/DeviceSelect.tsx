import {
  Select,
  SelectContent,
  SelectGroup,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from "@src/components/ui/select";

export type SelectProps = {
  onValueChange: (value: string) => void;
  value: string | number | boolean | null | undefined;
};

type DeviceSelectProps<T> = {
  onValueChange: (value: string) => void;
  value: string | number | boolean | null | undefined;
  placeholder?: string;
  devices: T[] | undefined;
  keySelector: (device: T) => string;
  valueSelector: (device: T) => string;
  nameSelector: (device: T) => string;
};

export const DeviceSelect = <T,>({
  onValueChange,
  value,
  placeholder,
  devices,
  keySelector,
  valueSelector,
  nameSelector,
}: DeviceSelectProps<T>) => {
  const found = devices && devices.length > 0;

  const selectedDevice = devices?.find((d) => valueSelector(d) === value);
  const selectedDeviceName = selectedDevice
    ? nameSelector(selectedDevice)
    : undefined;

  return (
    <Select onValueChange={onValueChange}>
      <SelectTrigger
        className="border-none bg-background focus:ring-accent1 focus:ring-offset-1 focus-visible:ring-accent1 focus-visible:ring-offset-1 "
        disabled={!found}
      >
        <SelectValue
          placeholder={
            !selectedDeviceName && !found ? placeholder : selectedDeviceName
          }
        />
      </SelectTrigger>
      <SelectContent className="max-h-72">
        <SelectGroup>
          {found &&
            devices.map((d) => (
              <SelectItem key={keySelector(d)} value={valueSelector(d)}>
                {nameSelector(d)}
              </SelectItem>
            ))}
        </SelectGroup>
      </SelectContent>
    </Select>
  );
};
