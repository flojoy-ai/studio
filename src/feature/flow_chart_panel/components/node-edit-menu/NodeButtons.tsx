import { Box, createStyles } from "@mantine/core";
import { IconArrowsMaximize, IconX } from "@tabler/icons-react";
import { SetStateAction } from "jotai";
import { Dispatch } from "react";
import { ElementsData } from "../../types/CustomNodeProps";

const useStyles = createStyles((theme) => ({
  Edit: {
    position: "absolute",
    padding: 4,
    width: "100%",
    zIndex: 150,
    display: "flex",
    color:
      theme.colorScheme === "dark"
        ? theme.colors.gray[1]
        : theme.colors.dark[8],
  },
}));

type NodeButtonsProps = {
  data: ElementsData;
  setIsExpandMode: Dispatch<SetStateAction<boolean>>;
};

const NodeButtons = ({ data, setIsExpandMode }: NodeButtonsProps) => {
  const { classes } = useStyles();

  const onNodeExpandClick = () => {
    setIsExpandMode(true);
  };

  const handleXButtonClick = () => {
    if (!data.handleRemove) {
      console.error("NodeButtons: handleRemove callback not attached");
      return;
    }
    data.handleRemove(data.id);
  };

  return (
    <Box className={classes.Edit}>
      <Box sx={{ cursor: "pointer" }}>
        <IconArrowsMaximize
          data-testid="expand-button"
          onClick={onNodeExpandClick}
        />
      </Box>
      <Box ml="auto" sx={{ cursor: "pointer" }}>
        <IconX onClick={handleXButtonClick} />
      </Box>
    </Box>
  );
};

export default NodeButtons;
