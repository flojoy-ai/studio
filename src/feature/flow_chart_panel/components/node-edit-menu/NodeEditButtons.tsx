import { Box, createStyles } from "@mantine/core";
import { useFlowChartState } from "@src/hooks/useFlowChartState";
import { IconArrowsMaximize, IconPencil, IconX } from "@tabler/icons-react";
import { ElementsData } from "../../types/CustomNodeProps";

const useStyles = createStyles((theme) => ({
  Edit: {
    position: "absolute",
    top: 7,
    right: 3,
    zIndex: 150,
    display: "flex",
    cursor: "pointer",
    color:
      theme.colorScheme === "dark"
        ? theme.colors.gray[1]
        : theme.colors.dark[8],
  },
  leftEdit: {
    position: "absolute",
    top: 8,
    left: 7,
    zIndex: 150,
    display: "flex",
    cursor: "pointer",
    color:
      theme.colorScheme === "dark"
        ? theme.colors.gray[1]
        : theme.colors.dark[8],
  },
}));

type NodeEditButtonsProps = {
  data: ElementsData;
  handleRemove: (id: string) => void;
  showPencil: boolean;
};

const NodeEditButtons = ({
  data,
  handleRemove,
  showPencil,
}: NodeEditButtonsProps) => {
  const { setIsEditMode, setIsExpandMode } = useFlowChartState();
  const { classes } = useStyles();

  const onNodeExpandClick = () => {
    setIsExpandMode(true);
  };

  return (
    <div>
      <Box className={classes.leftEdit}>
        <IconArrowsMaximize onClick={onNodeExpandClick} />
        {showPencil && <IconPencil onClick={() => setIsEditMode(true)} />}
      </Box>
      <Box className={classes.Edit}>
        <IconX onClick={() => handleRemove(data.id)} />
      </Box>
    </div>
  );
};

export default NodeEditButtons;
