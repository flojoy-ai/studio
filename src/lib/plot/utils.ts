export function clamp(x: number, lo: number, hi: number) {
  return Math.min(Math.max(x, lo), hi);
}
