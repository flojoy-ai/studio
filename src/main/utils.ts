import * as http from "http";
import * as fs from "fs";
import { isIP } from "net";
import { execCommand } from "./executor";
import { Command } from "./command";
import { app, dialog } from "electron";
import { join } from "path";
import { killCaptain } from "./python";
import log from "electron-log";
import { ChildProcess } from "node:child_process";
import { Err, Result, Ok } from "../types/result";

export const isPortFree = (port: number) =>
  new Promise((resolve) => {
    const server = http
      .createServer()
      .listen(port, "127.0.0.1", () => {
        server.close();
        resolve(true);
      })
      .on("error", () => {
        resolve(false);
      });
  });

export const killProcess = async (port: number) => {
  await execCommand(
    new Command({
      darwin: `kill -9 $(lsof -t -i :${port})`,
      linux: `kill -9 $(lsof -t -i :${port})`,
      win32: `FOR /F "tokens=5" %i IN ('netstat -aon ^| find "${port}"') DO Taskkill /F /PID %i`,
    }),
  );
};

export const ping = async (addr: string) => {
  if (isIP(addr) === 0) {
    throw new Error(`Invalid IP address: ${addr}`);
  }

  return await execCommand(
    new Command({
      win32: `ping -n 1 ${addr}`,
      darwin: `ping -c 1 ${addr}`,
      linux: `ping -c 1 ${addr}`,
    }),
    { quiet: true },
  );
};

export const netstat = async () => {
  return await execCommand(
    new Command({
      win32: "netstat",
      darwin: "netstat",
      linux: "netstat",
    }),
    { quiet: true },
  );
};

export const ifconfig = async () => {
  return await execCommand(
    new Command({
      win32: "ipconfig /all",
      darwin: "ifconfig",
      linux: "ifconfig",
    }),
    { quiet: true },
  );
};

export const writeFileSync = (_, filePath: string, text: string): void => {
  fs.writeFileSync(filePath, text);
};

export const pickDirectory = async (): Promise<string> => {
  const handler = await dialog.showOpenDialog({
    properties: ["openDirectory"],
  });
  return handler.canceled ? "" : handler.filePaths[0];
};

export const getCustomBlocksDir = async () => {
  const filePath = join(app.getPath("home"), ".flojoy/custom_blocks_path.txt");

  if (!fs.existsSync(filePath)) {
    return undefined;
  }

  const blocksPath = fs
    .readFileSync(filePath, { encoding: "utf-8" })
    .toString();

  if (!fs.existsSync(blocksPath)) {
    return undefined;
  }

  return blocksPath;
};

export const cacheCustomBlocksDir = (_, dirPath: string) => {
  const flojoyDir = join(app.getPath("home"), ".flojoy");
  if (!fs.existsSync(flojoyDir)) {
    fs.mkdirSync(flojoyDir);
  }
  const cacheFilePath = join(flojoyDir, "custom_blocks_path.txt");
  fs.writeFileSync(cacheFilePath, dirPath);
};

export const openTestPicker = (): Promise<
  { filePath: string; fileContent: string } | undefined
> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: ["json", "py"],
            name: "Test",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const openFilePicker = (
  _,
  allowedExtensions: string[] = ["json"],
): Promise<{ filePath: string; fileContent: string } | undefined> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: allowedExtensions,
            name: "allowed extensions",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
        resolve(undefined);
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const cleanup = async () => {
  const captainProcess = global.captainProcess as ChildProcess;
  log.info(
    "Cleanup function invoked, is captain running? ",
    !(captainProcess?.killed ?? true),
  );
  if (captainProcess && captainProcess.exitCode === null) {
    const success = killCaptain();
    if (success) {
      global.captainProcess = null;
      log.info("Successfully terminated captain :)");
    } else {
      log.error("Something went wrong when terminating captain!");
    }
  }
};

export const loadFileFromFullPath = async (
  filePath: string,
): Promise<string> => {
  return fs.readFileSync(filePath, { encoding: "utf-8" }).toString();
};

export const saveFileToFullPath = async (
  filePath: string,
  content: string,
): Promise<Result<void>> => {
  try {
    fs.writeFileSync(filePath, content);
    return Ok(undefined);
  } catch (error) {
    return Err(error as Error);
  }
};
