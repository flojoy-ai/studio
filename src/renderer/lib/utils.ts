import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";
import { z } from "zod";
import * as E from "fp-ts/Either";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function tryParse<Z extends z.ZodTypeAny>(z: Z) {
  return (val: unknown) => {
    const res = z.safeParse(val);
    if (res.success) {
      return E.right(res.data as z.infer<Z>);
    }
    return E.left(res.error);
  };
}
