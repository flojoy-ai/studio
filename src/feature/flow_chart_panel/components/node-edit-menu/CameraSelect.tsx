import { useHardwareDevices } from "@src/hooks/useHardwareDevices";
import { DeviceSelect, SelectProps } from "./DeviceSelect";

export const CameraSelect = (props: SelectProps) => {
  const hardware = useHardwareDevices();
  const cameras = hardware?.cameras;

  return (
    <DeviceSelect
      {...props}
      devices={cameras}
      placeholder="No cameras found"
      keySelector={(cam) => cam.id.toString()}
      valueSelector={(cam) => cam.id.toString()}
      nameSelector={(cam) => cam.name}
    />
  );
};
