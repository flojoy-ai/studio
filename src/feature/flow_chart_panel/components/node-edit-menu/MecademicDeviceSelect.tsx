import { useHardwareDevices } from "@src/hooks/useHardwareDevices";
import { DeviceSelect, SelectProps } from "./DeviceSelect";

export const MecademicDeviceSelect = (props: SelectProps) => {
  const hardware = useHardwareDevices();
  const mecas = hardware?.mecademicDevices;

  return (
    <DeviceSelect
      {...props}
      devices={mecas}
      placeholder="No cameras found"
      keySelector={(meca) => meca.ip.toString()}
      valueSelector={(meca) => meca.ip.toString()}
      nameSelector={(meca) => meca.name}
    />
  );
};
