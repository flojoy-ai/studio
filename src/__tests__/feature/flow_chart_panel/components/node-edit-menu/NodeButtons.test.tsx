import { renderWithTheme } from "@src/__tests__/__utils__/utils";
import NodeButtons from "@src/feature/flow_chart_panel/components/node-edit-menu/NodeButtons";
import { fireEvent } from "@testing-library/dom";
import { useState } from "react";

const mockData = {
  id: "test-id",
  label: "test",
  func: "test",
  type: "test",
  ctrls: {},
  path: "",
};

const handleRemove = jest.fn();

jest.doMock(
  "@src/feature/flow_chart_panel/components/node-edit-menu/NodeButtons",
  () => {
    const NodeButtonsMock = () => {
      const [modalOpen, setModalOpen] = useState(false);
      return (
        <>
          {modalOpen && <div data-testid="node-modal" />}
          <NodeButtons
            handleRemove={handleRemove}
            data={mockData}
            setIsExpandMode={setModalOpen}
          />
        </>
      );
    };
    return { NodeButtons: NodeButtonsMock };
  }
);

const NodeButtonsTest =
  require("@src/feature/flow_chart_panel/components/node-edit-menu/NodeButtons").NodeButtons;

describe("NodeButtons component", () => {
  it("renders correctly with default props", () => {
    const { container } = renderWithTheme(
      <NodeButtons
        handleRemove={handleRemove}
        data={mockData}
        setIsExpandMode={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("opens modal when expand button is clicked", () => {
    const { getByTestId } = renderWithTheme(<NodeButtonsTest />);
    const expandButton = getByTestId("expand-button");

    fireEvent.click(expandButton);

    const modal = getByTestId("node-modal");
    expect(modal).toBeInTheDocument();
  });
});
