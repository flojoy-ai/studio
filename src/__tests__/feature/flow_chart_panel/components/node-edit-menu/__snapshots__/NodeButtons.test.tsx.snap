// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeButtons component renders correctly with default props 1`] = `
<div>
  <div
    class="mantine-1adffj6"
  >
    <div
      class="mantine-ri1b3q"
    >
      <svg
        class="tabler-icon tabler-icon-arrows-maximize"
        data-testid="expand-button"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 4l4 0l0 4"
        />
        <path
          d="M14 10l6 -6"
        />
        <path
          d="M8 20l-4 0l0 -4"
        />
        <path
          d="M4 20l6 -6"
        />
        <path
          d="M16 20l4 0l0 -4"
        />
        <path
          d="M14 14l6 6"
        />
        <path
          d="M8 4l-4 0l0 4"
        />
        <path
          d="M4 4l6 6"
        />
      </svg>
    </div>
  </div>
  ,
</div>
`;
